pkg/webhook/webhook.go

```diff
@@ -0,0 +1,345 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package webhook
+
+import (
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"github.com/mattbaird/jsonpatch"
+	//"github.com/pkg/errors"
+	//"github.com/sirupsen/logrus"
+	"github.com/golang/glog"
+
+	admv1beta1 "k8s.io/api/admission/v1beta1"
+	admregv1beta1 "k8s.io/api/admissionregistration/v1beta1"
+	"k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	kubeclientset "k8s.io/client-go/kubernetes"
+
+	snapshotv1alpha1 "github.com/kubernetes-csi/external-snapshotter/pkg/apis/volumesnapshot/v1alpha1"
+	clientset "github.com/kubernetes-csi/external-snapshotter/pkg/client/clientset/versioned"
+)
+
+const (
+	webhookConfigName   = "snapshot-webhook-config"
+	snapshotWebhookName = "snapshot-init.snapshot.storage.k8s.com"
+	// The admission webhook must be exposed in the following service,
+	// this is mainly for the server certificate.
+	// serviceName is the name of the admission webhook service.
+	//serviceName = "snapshot-webhook-svc"
+	serviceName = "snapshot-admission-webhook-svc"
+
+	IsDefaultSnapshotClassAnnotation = "snapshot.storage.kubernetes.io/is-default-class"
+)
+
+var supportedSourceKinds = sets.NewString(string("PersistentVolumeClaim"))
+var supportedDataSourceAPIGroups = sets.NewString(string(""))
+
+// CreateConfiguration creates MutatingWebhookConfiguration and registeres the
+// webhood admission controller with the kube-apiserver
+func CreateConfiguration(clientset kubeclientset.Interface, url string) error {
+	failurePolicy := admregv1beta1.Fail
+	config := &admregv1beta1.MutatingWebhookConfiguration{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: webhookConfigName,
+		},
+		Webhooks: []admregv1beta1.Webhook{
+			// Webhook for initializing Pods.
+			{
+				Name: snapshotWebhookName,
+				ClientConfig: admregv1beta1.WebhookClientConfig{
+					Service: &admregv1beta1.ServiceReference{
+						Name:      serviceName,
+						Namespace: "default",
+					},
+					CABundle: caCert,
+				},
+				Rules: []admregv1beta1.RuleWithOperations{
+					{
+						Operations: []admregv1beta1.OperationType{
+							admregv1beta1.Create,
+						},
+						Rule: admregv1beta1.Rule{
+							APIGroups:   []string{snapshotv1alpha1.GroupName},
+							APIVersions: []string{"v1alpha1"},
+							Resources:   []string{"volumesnapshots"},
+						},
+					},
+				},
+				FailurePolicy: &failurePolicy,
+			},
+		},
+	}
+	glog.V(2).Infof("Creating ValidatingWebhookConfiguration %q", config.Name)
```

Comment:
Should this be a V(5) for debug msg?

Reply:
I think it is ok since it only shows up once.